File: Valt/Compiler/Typing/ResolvedType.cs
using Microsoft.VisualBasic;
using Valt.Compiler.Declarations;

namespace Valt.Compiler.Typing
{
    public enum ValtInternalType
    {
        Primitive = 0,
        CType,
        Array,
        Struct,
        Enum,
        Map
    }
    public class ResolvedType
    {
        public string Name;
        public ResolvedTypeKind Kind { get; set; }
        public ResolvedType ElementType { get; set; }
        public bool IsPointer { get; set; }
        public bool IsReference { get; set; }
        public NamedDeclaration DataRef { get; set; }

        public int FixedArrayLength { get; set; } = -1;

        public ValtInternalType InternalType { get; set; }
        public override string ToString()
        {
            if (DataRef != null)
                return DataRef.ToString();
            string formatData = "";
            if (IsPointer)
            {
                formatData = "*";
            }
            if (IsReference)
            {
                formatData = "&";
            }
            if (InternalType==ValtInternalType.Array)
            {
                formatData = "[]";
            }
            return $"{Kind} {formatData}{Name}";
        }

    }
}

File: JavaPort/src/typing/ResolvedType.kt
package typing

import declarations.NamedDeclaration

enum class ValtInternalType
{
    Primitive,
    CType,
    Array,
    Struct,
    Enum,
    Map
}

enum class ResolvedTypeKind {
    Primitive, Struct, Enum, Union, Map
}
class ResolvedType
{
    var Name = ""
    var Kind = ResolvedTypeKind.Primitive
    var ElementType : ResolvedType ? = null
    var ValueType: ResolvedType? = null
    var IsPointer = false
    var IsReference = false
    var DataRef=  NamedDeclaration()

    var FixedArrayLength = -1

    var InternalType = ValtInternalType.Enum
}